### ISSUE TITLE
1.8.2 and es6.jsx

### ISSUE THREAD

**Issue by ngan** [reactjs/react-rails#587]
Help us help you! Please choose one:
- [x] My app crashes with `react-rails`, so I've included the stack trace and the exact steps which make it crash.
- [ ] My app doesn't crash, but I'm getting unexpected behavior. So, I've described the unexpected behavior and suggested a new behavior.
- [ ] I'm trying to use `react-rails` with another library, but I'm having trouble. I've described my JavaScript management setup (eg, Sprockets, Webpack...), how I'm trying to use this other library, and why it's not working.
- [ ] I have another issue to discuss.

---

Upgraded to 1.8.2 and started seeing this error:

```
Sprockets::FileNotFound - couldn't find file 'classnames' with type 'application/jsx'
Checked in these paths: 
```

I'm using the `rails-assets-classnames` gem, but I think the problem is because it's looking for `classnames.js.jsx` file. My component file is `my_component.es6.jsx`.

Downgraded to 1.8.1 and things worked again.

**Comment by rmosolgo** [comment]
Could you share your Sprockets version from `Gemfile.lock`? It seems like we have issues with 3.7+

https://github.com/reactjs/react-rails/pull/582

https://github.com/reactjs/react-rails/pull/574

cc/ @seuros

**Comment by seuros** [comment]
@ngan provide us with a sample app that reprocude this bug, and i will try to fix it.

**Comment by seuros** [comment]
1.8.1 is defective (don't process directives).

**Comment by tirdadc** [comment]
Exact same issue, prerender works fine with 1.8.1 (or at least didn't cause things to crash) and sprockets 3.7.0, I upgrade to 1.8.2 and I get this:

```
Sprockets::FileNotFound - couldn't find file 'moment' with type 'application/jsx'
Checked in these paths:
  /private/var/srv/otioso/app/assets/images
  /private/var/srv/otioso/app/assets/javascripts
  /private/var/srv/otioso/app/assets/stylesheets
  /private/var/srv/otioso/vendor/assets/javascripts
  /private/var/srv/otioso/vendor/assets/stylesheets
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/lodash-rails-4.14.1/vendor/assets/javascripts
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/momentjs-rails-2.11.1/vendor/assets/javascripts
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/coffee-rails-4.2.1/lib/assets/javascripts
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/jquery-ui-rails-5.0.5/app/assets/images
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/jquery-ui-rails-5.0.5/app/assets/javascripts
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/jquery-ui-rails-5.0.5/app/assets/stylesheets
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/jquery-rails-4.2.1/vendor/assets/javascripts
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/turbolinks-source-5.0.0/lib/assets/javascripts
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/react-rails-1.8.2/lib/assets/react-source/development-with-addons
  /Users/tirdad/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/react-rails-1.8.2/lib/assets/javascripts:
```

this is my `server-rendering.js`:

```
//= require moment.js
//= require lodash.min
//= require react-server
//= require components
```

and I have this line in `config/initializers/assets.rb`:
```
Rails.application.config.assets.precompile += %w( vendor.js vendor.css server_rendering.js )
```

Exact details for Sprockets in my Gemfile.lock:
```
    sprockets (3.7.0)
      concurrent-ruby (~> 1.0)
      rack (> 1, < 3)
    sprockets-rails (3.1.1)
      actionpack (>= 4.0)
      activesupport (>= 4.0)
      sprockets (>= 3.0.0)
```

**Comment by rmosolgo** [comment]
Thanks for the details, I'll take a look soon!

**Comment by seuros** [comment]
@tirdadc could you help us by providing a small sample app that crash with this bug ? It will speedup thing a log. Thanks you.

**Comment by rmosolgo** [comment]
Added a test case here, maybe something will turn up: https://github.com/reactjs/react-rails/pull/591

**Comment by rmosolgo** [comment]
Seems like we're getting `application/jsx` here when we _want_ `application/javascript`

https://github.com/rails/sprockets/blob/3.x/lib/sprockets/directive_processor.rb#L76

**Comment by rmosolgo** [comment]
Coffeescript is a similar setup, and it works fine with Sprockets 3.7. Perhaps it's because they're using the old interface (`register_engine`) and silencing deprecation: 

https://github.com/rails/sprockets/blob/3.x/lib/sprockets.rb#L124-L128

I'm beat on how to "solve" this issue (ie, how to use `register_processor` with Sprockets 3.7+). I think it requires either: 

- some way for JSX to require dependencies as `application/javascript`, not `application/jsx`; OR 
- some way to register transformers from `application/javascript` to `application/jsx`. I tried: 

  ```ruby 
  null_transformer = -> (input) { input[:data] }
  sprockets_env.register_transformer("application/javascript", "application/jsx", null_transformer)
  ```

  but it didn't seem to work, and I'm not sure how that would handle CoffeeScript, since it doesn't have its own mime type in 3.7 

My suggestion is to revert to the previous behavior (`register_processor` for 4+ only) and silence deprecations (as suggested in #572 , implemented again here: fdfec81d87b0370291d1ee4c325109b055a45cfb)

**Comment by eysi09** [comment]
I'm also getting the `application/jsx` error with react-rails 1.8.2 and sprockets 3.7.0. A comment in the docs would be helpful, it took me awhile to figure out that downgrading to 1.8.1 would do the trick.

**Comment by calebwoods** [comment]
This is fixed for me using the latest on `master` would be great to get a new gem version with the fix

**Comment by rmosolgo** [comment]
Just shipped 1.9.0 with that fix, thanks for trying it out!
